packages/components/src/detect-focus-outside/README.md

```diff
@@ -0,0 +1,36 @@
+# DetectFocusOutside
```

Comment:
Lately, I have been wondering if it's better to avoid components for these "behavior-based" components and instead write hooks that accept "refs" or return elements (like useResizeAware). The advantage being to avoid the extra wrapper.

Reply:
Hm, I'm not sure exactly how that would look in this case. `useResizeAware` returns an element which is expected to be rendered as a child of a component. We need to capture events bubbling from within, so it would at least need to be a _wrapping_ element, which would basically be the same as what this implementation is already.

I do think it would be ideal if we could at least _flatten_ the DOM hierarchy, since currently this component must add a `div` wrapper. We don't strictly need the `div`, aside from being able to capture bubbling events. I recall having some hope that in the future, React might support event handlers to be assigned to `Fragment`, since their synthetic events don't use the DOM anyways and it can technically be possible.

In fact, they still note this in their documentation:

>`key` is the only attribute that can be passed to `Fragment`. In the future, we may add support for additional attributes, such as event handlers.

https://reactjs.org/docs/fragments.html#keyed-fragments

I think I saw later that they might be avoiding that option though (perhaps part of their new ["Fire" effort](https://github.com/facebook/react/issues/13525)).

So maybe it's not worth placing our bets on this. The alternatives might have some concerning implication on the ergonomics of the component:

- We could have a hook which returns an object of props for the event handlers. But how do we handle a case where both the hook and the original component each need to have an e.g. `onFocus` event handler?
- We could have a hook which returns a `ref`, and then the hook attaches DOM event handlers to the `ref` instance. But should we have any concern that those are _DOM_ events, not React events, which behave (and bubble) slightly differently? We've had issues with this in the past for things like `Slot`/`Fill` `bubblesVirtually`.

One possible _advantage_ of using the DOM events is that we could use the actual [`blur` event](https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event), where the distinction between `blur` and [`focusout`](https://developer.mozilla.org/en-US/docs/Web/API/Element/focusout_event) is basically the point of what this component is trying to implement. As far as I can recall, React counter-intuitively uses the `focusout` event when assigning `onBlur` (see https://github.com/facebook/react/issues/6410, specifically https://github.com/facebook/react/issues/6410#issuecomment-207064994).

Reply:
I'm not sure if a good pattern or not but on previous occasions I've wrote hooks like that:

```
useSomething = ( ref ) => {
   useEffect(() => {
     // do something with ref.current.
     // useEffect ensures the ref is defined
  });
}
```

Reply:
I believe it would be good if we find a pattern here and follow it for future hooks relying on refs.

Reply:
I think the harder question is if and when we "rely on refs", including whether it makes sense here.